src/query/server/query.go

```diff
@@ -643,18 +644,35 @@ func Run(runOpts RunOptions) RunResult {
 		}
 	}
 
-	prometheusEngine, err := newPromQLEngine(cfg, prometheusEngineRegistry,
-		instrumentOptions)
+	prometheusEngineFn := func(lookbackDuration time.Duration) (*prometheuspromql.Engine, error) {
+		// NB: use nil metric registry to avoid duplicate metric registration when creating multiple engines
+		return newPromQLEngine(lookbackDuration, cfg, nil, instrumentOptions)
+	}
+	engineCache := promqlengine.NewCache(prometheusEngineFn)
+	if mult := cfg.Middleware.Prometheus.ResolutionMultiplier; mult > 0 {
+		for _, res := range extractNamespaceResolutions(runOpts.Config.Clusters) {
+			if res > 0 {
+				resolutionBasedLookback := res * time.Duration(mult)
+				eng, err := prometheusEngineFn(resolutionBasedLookback)
+				if err != nil {
+					logger.Fatal("unable to create PromQL engine", zap.Error(err))
+				}
+				engineCache.Set(resolutionBasedLookback, eng)
+			}
+		}
+	}
+	defaultPrometheusEngine, err := newPromQLEngine(lookbackDuration, cfg, prometheusEngineRegistry, instrumentOptions)
```

Comment:
I understand that the intention here is to overwrite the `defaultPrometheusEngine` in the cache (in case it was already added for the same duration in the loop above). But this feels a bit obscure. I would move creation of the default engine to the beginning, and add a check against duplication in the resolutions loop.

Reply:
https://github.com/m3db/m3/pull/3882/commits/d22380bcc23113f4b86226962dc2a065357db3cf#diff-500287f1f210c991514dddeaf205ae4b641f2fb9a3173b7470b4148f180c9893L651-R674